rustls-libssl/src/sign.rs

```diff
@@ -0,0 +1,211 @@
+    }
+}
+
+#[derive(Debug)]
+struct OpenSslKey(EvpPkey);
+
+impl sign::SigningKey for OpenSslKey {
+    fn choose_scheme(&self, offered: &[SignatureScheme]) -> Option<Box<dyn sign::Signer>> {
+        match self.0.algorithm() {
+            SignatureAlgorithm::RSA => {
+                if offered.contains(&SignatureScheme::RSA_PSS_SHA512) {
+                    return Some(Box::new(OpenSslSigner {
+                        pkey: self.0.clone(),
+                        pscheme: rsa_pss_sha512(),
+                        scheme: SignatureScheme::RSA_PSS_SHA512,
+                    }));
+                }
+                if offered.contains(&SignatureScheme::RSA_PSS_SHA384) {
+                    return Some(Box::new(OpenSslSigner {
+                        pkey: self.0.clone(),
+                        pscheme: rsa_pss_sha384(),
+                        scheme: SignatureScheme::RSA_PSS_SHA384,
+                    }));
+                }
+                if offered.contains(&SignatureScheme::RSA_PSS_SHA256) {
+                    return Some(Box::new(OpenSslSigner {
+                        pkey: self.0.clone(),
+                        pscheme: rsa_pss_sha256(),
+                        scheme: SignatureScheme::RSA_PSS_SHA256,
+                    }));
+                }
+
+                if offered.contains(&SignatureScheme::RSA_PKCS1_SHA512) {
+                    return Some(Box::new(OpenSslSigner {
+                        pkey: self.0.clone(),
+                        pscheme: rsa_pkcs1_sha512(),
+                        scheme: SignatureScheme::RSA_PKCS1_SHA512,
+                    }));
+                }
+                if offered.contains(&SignatureScheme::RSA_PKCS1_SHA384) {
+                    return Some(Box::new(OpenSslSigner {
+                        pkey: self.0.clone(),
+                        pscheme: rsa_pkcs1_sha384(),
+                        scheme: SignatureScheme::RSA_PKCS1_SHA384,
+                    }));
+                }
+                if offered.contains(&SignatureScheme::RSA_PKCS1_SHA256) {
+                    return Some(Box::new(OpenSslSigner {
+                        pkey: self.0.clone(),
+                        pscheme: rsa_pkcs1_sha256(),
+                        scheme: SignatureScheme::RSA_PKCS1_SHA256,
+                    }));
+                }
+
+                None
+            }
+            _ => None,
+        }
+    }
+
+    fn algorithm(&self) -> SignatureAlgorithm {
+        self.0.algorithm()
+    }
+}
+
+#[derive(Debug)]
+struct OpenSslSigner {
+    pkey: EvpPkey,
+    pscheme: Box<dyn EvpScheme + Send + Sync>,
+    scheme: SignatureScheme,
+}
+
+impl sign::Signer for OpenSslSigner {
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, rustls::Error> {
+        self.pkey
+            .sign(self.pscheme.as_ref(), message)
+            .map_err(|_| rustls::Error::General("signing failed".to_string()))
+    }
+
+    fn scheme(&self) -> SignatureScheme {
+        self.scheme
+    }
```

Comment:
could this be replaced by just making the `scheme` field `pub`?

Reply:
It's a function item required by `impl sign::Signer`

Reply:
Ohh of course. Thanks